File: 2025-2026-Spring/Lecture_1/pi_calculation.py
import random, time

def estimate_pi_sequential(n_points):
    inside = 0
    for _ in range(n_points):
        x, y = random.random(), random.random()
        if x*x + y*y <= 1.0:
            inside += 1
    return 4.0 * inside / n_points

N = 10_000_000
start = time.perf_counter()
pi = estimate_pi_sequential(N)
elapsed = time.perf_counter() - start
print(f"π ≈ {pi:.6f}  |  time: {elapsed:.3f}s")

File: 2025-2026-Spring/Lecture_1/pi_calculation.cpp
// C++ — Sequential Pi: compiled baseline
// ~10-20x faster than Python: compiled, typed, no interpreter overhead
#include <cstdlib>
#include <chrono>
#include <cstdio>
#include <random>

double estimate_pi(long n) {
    std::mt19937_64 rng(std::random_device{}());
    std::uniform_real_distribution<double> dist(0.0, 1.0);

    long inside = 0;
    for (long i = 0; i < n; i++) {
        double x = dist(rng);
        double y = dist(rng);
        if (x*x + y*y <= 1.0) inside++;
    }
    return 4.0 * inside / n;
}

int main() {
    const long N = 10'000'000;

    auto start = std::chrono::high_resolution_clock::now();
    double pi = estimate_pi(N);
    auto end   = std::chrono::high_resolution_clock::now();

    double elapsed = std::chrono::duration<double>(end - start).count();
    std::printf("Pi ≈ %.6f  |  n = %ld  |  time = %.3fs\n", pi, N, elapsed);
    return 0;
}
